File: src/behavioral/visitor2.java
@Data
public abstract class Client {
    private final String name;
    private final String address;
    private final String number;

    public abstract void sendMail();
}

@Getter
public class Resident extends Client {
    @Override
    public void sendMail() {
        System.out.println("Sending mail about medical insurance...");
    }
}

@Getter
public class Company extends Client {
    @Override
    public void sendMail() {
        System.out.println("Sending employees and equipment insurance mail...");
    }
}

@Getter
public class Bank extends Client {
    @Override
    public void sendMail() {
        System.out.println("Sending mail about theft insurance...");
    }
}

@Getter
public class Restaurant extends Client {
    @Override
    public void sendMail() {
        System.out.println("Sending mail about fire and food insurance...");
    }
}

File: src/behavioral/visitor2.ts
abstract class Client {
  constructor(
    protected name: string,
    protected address: string,
    protected number: string,
  ) {}

  abstract sendMail(): void;
}

export class Resident extends Client {
  constructor(
    name: string,
    address: string,
    number: string,
    readonly insuranceClass: string,
  ) {
    super(name, address, number);
    this.insuranceClass = insuranceClass;
  }

  getName(): string {
    return this.name;
  }

  sendMail(): void {
    console.log("Sending mail about medical insurance...");
  }
}

export class Company extends Client {
  constructor(
    name: string,
    address: string,
    number: string,
    readonly nbrOfEmployees: number,
  ) {
    super(name, address, number);
    this.nbrOfEmployees = nbrOfEmployees;
  }

  sendMail(): void {
    console.log("Sending employees and equipment insurance mail...");
  }
}

export class Bank extends Client {
  constructor(
    name: string,
    address: string,
    number: string,
    readonly branchesInsured: number,
  ) {
    super(name, address, number);
    this.branchesInsured = branchesInsured;
  }

  sendMail(): void {
    console.log("Sending mail about theft insurance...");
  }
}

export class Restaurant extends Client {
  constructor(
    name: string,
    address: string,
    number: string,
    readonly availableAbroad: boolean,
  ) {
    super(name, address, number);
    this.availableAbroad = availableAbroad;
  }

  sendMail(): void {
    console.log("Sending mail about fire and food insurance...");
  }
}
